rfc-120-csv-preview.md

```diff
@@ -0,0 +1,134 @@
+
+- Using a separate content item means we're not coupled to a particular rendering app. This provides a foundation for extending the preview feature in the future.
+
+### Step 2
+
+Re-implement the CSV Preview feature in Government Frontend.
+
+- Government Frontend currently renders all documents published by Whitehall, that have CSV file attachments. This makes it a relevant codebase to implement this feature in.
+
+- We expect the implementation to be small, isolated and not to grow/change much. Adding a new responsibility to Government Frontend seems like a reasonable compromise when compared to the [rejected proposals](#rejected-proposals).
+
+
+### Step 3
+
+Change Whitehall to use the new CSV Preview feture for newly published documents. We will need to use a different URL structure, since the current one is coupled to Whitehall.
+
+```
+# current URL example
+https://www.gov.uk/government/uploads/system/uploads/attachment_data/file/856932/Management_information_-_further_education_and_skills_-_as_at_31_December_2019.csv/preview
+
+# proposed URL example
+https://www.gov.uk/government/remaining-parent-document-base-path/preview/asset-id
+```
+
+The `asset-id` should be the ID used by Asset Manager e.g. `597b098a759b743e0b759a96`, since it persists for the lifetime of the asset, even if the original file is replaced.
+
+
+### Step 4
+
+Change existing attachments (in Whitehall) to use the new CSV Preview. This will involve republishing all Whitehall documents that include a CSV preview link.
+
+Once this is done, we can remove some of the config for [the `assets` domain](https://github.com/alphagov/govuk-puppet/blob/51d7f7b648e8fde5aae3adfb774ec3bca6325bd8/modules/router/templates/assets_origin.conf.erb#L69-L74) that supported the legacy feature ([`gov.uk` routing](https://github.com/alphagov/government-frontend/blob/dcbc563030c470e6242d6432896d84feebb2138b/config/routes.rb#L14) remains). We can also delete the code for the legacy feature.
+
+> We may want to add redirects for the old preview URLs in Whitehall, if they continue to receive lots of traffic after all the 'View Online' links have updated.
+
+
+## Rejected proposals
+
+### Use a top-level prefix for the routing (e.g. `/preview/:base_path/:attachment_id`)
+
+This would involve registering a new `/preview` prefix route in [router][router], which would set the backend (rendering app) to Government Frontend for all paths beginning with that route.
+
+- This involves less effort than individually managing lower-level routes for each attachment preview. However, because we don't have any knowledge of these routes, we can't help the user if the URL they enter is invalid; this could happen when the asset is removed, or the parent document changes e.g. unpublished, redirected.
```

Comment:
Could you elaborate on how/why the current proposal _would_ allow us to help the user if the URL they enter is invalid?

Reply:
There's some discussion of that with the current proposal [here](https://github.com/alphagov/govuk-rfcs/pull/122/files#diff-4e2dd4712e73fb35ede7ffcaf64a7e70R49):

> Using the publishing system means we can follow an existing approach for things like automatic redirects (in case the base path changes), unpublishing and anonymous preview ('fact check')

Reply:
I still don't really get it, apologies if I'm being slow!

Wouldn't we just have a route in Government Frontend like this:

```ruby
"/preview/:base_path/:attachment_id" => AttachmentHandler
```

...and in the `AttachmentHandler` code we'd make a request to Publishing API `https://publishing-api.dev.gov.uk/content/<attachment_id>` and either use the returned data or follow the redirect or display a "Withdrawn" page?